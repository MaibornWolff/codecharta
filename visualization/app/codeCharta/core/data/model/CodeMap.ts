export interface CodeMapNode {
    name: string,
    type: string,
    children?: CodeMapNode[]
    attributes: {
        [key: string]: number
    };
    deltas?: {
        [key: string]: number
    };
    link?: string;
    origin?: string;
    visible?: boolean;
    path?: string;
    markingColor?: string;
}

export interface CodeMap {

    fileName: string,
    apiVersion?: string,
    projectName: string,
    root: CodeMapNode,
    dependencies?: {
        static?: CodeMapDependency[],
        temporal_coupling?: CodeMapDependency[]
    }

}

export interface CodeMapDependency {
    node: string,
    nodeFilename?: string,
    dependantNode: string,
    dependantNodeFilename?: string,
    pairingRate?: number,
    averageRevs?: number
    visible: boolean
}