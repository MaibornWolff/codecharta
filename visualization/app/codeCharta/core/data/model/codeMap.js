export class CodeMap {

    constructor(fileName, projectName, root) {

        this.fileName = fileName;
        this.projectName = projectName;
        this.root = root;

    }


}