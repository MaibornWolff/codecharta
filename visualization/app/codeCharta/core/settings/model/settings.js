/**
 * This class defines the available set of settings in the CodeCharta application
 */
export class Settings {

    /**
     * @constructor
     * @param {CodeMap} map
     * @param {Range} neutralColorRange
     * @param {string} areaMetric
     * @param {string} heightMetric
     * @param {string} colorMetric
     * @param {boolean} deltas
     * @param {Number} amountOfTopLabels
     * @param {Scale} scaling
     * @param {Scale} camera
     * @param {Number} margin
     */
    constructor(map, neutralColorRange, areaMetric, heightMetric, colorMetric, deltas, amountOfTopLabels, scaling, camera, margin) {

        /**
         * currently selected map
         * @type {CodeMap}
         */
        this.map = map;

        /**
         * currently set neutral color range. This defines the yellow color range and the whether higher or lower values are green or red.
         * @type {Range}
         */
        this.neutralColorRange = neutralColorRange;

        /**
         * currently selected area metric
         * @type {string}
         */
        this.areaMetric = areaMetric;

        /**
         * currently selected heightm metric
         * @type {string}
         */
        this.heightMetric = heightMetric;

        /**
         * currently selected color metric
         * @type {string}
         */
        this.colorMetric = colorMetric;

        /**
         * delta flag
         * @type {boolean}
         */
        this.deltas = deltas;

        /**
         * number of highest buildings with labels
         * @type {Number}
         */
        this.amountOfTopLabels = amountOfTopLabels;

        /**
         * scaling settings
         * @type {Scale}
         */
        this.scaling = scaling;

        /**
         * camera position
         * @type {Scale}
         */
        this.camera = camera;

        /**
         * margin between buildings
         * @type {number}
         */
        this.margin = margin;

    }

    /**
     * Imports the given settings values without replacing this object with the given one.
     * @param {Settings} settings given settings
     */
    importSettingValues(settings){
        this.map = settings.map;
        this.neutralColorRange = settings.neutralColorRange;
        this.areaMetric = settings.areaMetric;
        this.heightMetric = settings.heightMetric;
        this.colorMetric = settings.colorMetric;
        this.deltas = settings.deltas;
        this.amountOfTopLabels = settings.amountOfTopLabels;
        this.scaling = settings.scaling;
        this.camera = settings.camera;
        this.margin = settings.margin;
    }

}
