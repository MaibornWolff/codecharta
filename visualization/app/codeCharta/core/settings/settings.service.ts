import {DataModel, DataService, DataServiceSubscriber} from "../data/data.service";
import {
    CameraChangeSubscriber,
    ThreeOrbitControlsService
} from "../../ui/codeMap/threeViewer/threeOrbitControlsService";
import {PerspectiveCamera} from "three";
import {STATISTIC_OPS} from "../statistic/statistic.service";
import {CodeMap, CodeMapNode, BlacklistItem, BlacklistType} from "../data/model/CodeMap";
import {hierarchy, HierarchyNode} from "d3-hierarchy";

export interface Range {
    from: number;
    to: number;
    flipped: boolean;
}

export interface Scale {
    x: number;
    y: number;
    z: number;
}

export enum KindOfMap {
    Single = "Single",
    Multiple = "Multiple",
    Delta = "Delta"
}

export interface Settings {

    map: CodeMap;
    neutralColorRange: Range;
    areaMetric: string;
    heightMetric: string;
    colorMetric: string;
    mode: KindOfMap;
    amountOfTopLabels: number;
    scaling: Scale;
    camera: Scale;
    margin: number;
    operation: STATISTIC_OPS;
    deltaColorFlipped: boolean;
    enableEdgeArrows: boolean;
    hideFlatBuildings: boolean;
    maximizeDetailPanel: boolean;
    invertHeight: boolean;
    dynamicMargin: boolean;
    isWhiteBackground: boolean;
    blacklist: Array<BlacklistItem>;
    focusedNodePath: string;
    searchedNodePaths: Array<string>;
    searchPattern: string;
}

export interface SettingsServiceSubscriber {
    onSettingsChanged(settings: Settings, event: Event);
}

export class SettingsService implements DataServiceSubscriber, CameraChangeSubscriber {

    public static SELECTOR = "settingsService";
    public static MIN_MARGIN = 15;
    public static MARGIN_FACTOR = 4;

    private _settings: Settings;

    public numberOfCalls: number;

    private _lastDeltaState = false;

    /* ngInject */
    constructor(private urlService, private dataService: DataService, private $rootScope,
                private threeOrbitControlsService: ThreeOrbitControlsService) {

        this._settings = this.getInitialSettings(dataService.data.renderMap, dataService.data.metrics);

        this.numberOfCalls = 0;
        dataService.subscribe(this);
        threeOrbitControlsService.subscribe(this);

    }

    public subscribe(subscriber: SettingsServiceSubscriber) {
        this.$rootScope.$on("settings-changed", (event, data) => {
            subscriber.onSettingsChanged(data, event);
        });
    }

    private getInitialSettings(renderMap: any, metrics: string[]): Settings {

        let r: Range = {
            from: 10,
            to: 20,
            flipped: false
        };

        let s: Scale = {
            x: 1, y: 1, z: 1
        };

        let c: Scale = {
            x: 0, y: 300, z: 1000
        };

        this._lastDeltaState = false;

        let settings: Settings = {
            map: renderMap,
            neutralColorRange: r,
            areaMetric: this.getMetricByIdOrLast(0, metrics),
            heightMetric: this.getMetricByIdOrLast(1, metrics),
            colorMetric: this.getMetricByIdOrLast(2, metrics),
            mode: KindOfMap.Single,
            amountOfTopLabels: 1,
            scaling: s,
            camera: c,
            margin: 15,
            operation: STATISTIC_OPS.NOTHING,
            deltaColorFlipped: false,
            enableEdgeArrows: true,
            hideFlatBuildings: true,
            maximizeDetailPanel: false,
            invertHeight: false,
            dynamicMargin: true,
            isWhiteBackground: false,
            blacklist: [],
            focusedNodePath: null,
            searchedNodePaths: [],
            searchPattern: null
        };
        return settings;

    }

    private onActivateDeltas() {
        this.dataService.onActivateDeltas();
    }

    private onDeactivateDeltas() {
        this.dataService.onDeactivateDeltas();
    }

    /**
     * change the map and metric settings according to the parameter.
     * @listens {data-changed} called on data-changed
     * @emits {settings-changed} transitively on call
     * @param {DataModel} data
     */
    public onDataChanged(data: DataModel) {

        if(data.metrics && data.renderMap && data.revisions) {
            this._settings.map = data.renderMap; // reference map is always the map which should be drawn
            this._settings.blacklist = data.renderMap.blacklist;

            if (data.metrics.indexOf(this._settings.areaMetric) === -1) {
                //area metric is not set or not in the new metrics and needs to be chosen
                this._settings.areaMetric = this.getMetricByIdOrLast(0, data.metrics);
            }

            if (data.metrics.indexOf(this._settings.heightMetric) === -1) {
                //height metric is not set or not in the new metrics and needs to be chosen
                this._settings.heightMetric = this.getMetricByIdOrLast(1, data.metrics);
            }

            if (data.metrics.indexOf(this._settings.colorMetric) === -1) {
                //color metric is not set or not in the new metrics and needs to be chosen
                this._settings.colorMetric = this.getMetricByIdOrLast(2, data.metrics);
            }

            this.onSettingsChanged();
        }

    }

    onCameraChanged(camera: PerspectiveCamera) {
        if (
            this._settings.camera.x !== camera.position.x ||
            this._settings.camera.y !== camera.position.y ||
            this._settings.camera.z !== camera.position.z
        ) {
            this._settings.camera.x = camera.position.x;
            this._settings.camera.y = camera.position.y;
            this._settings.camera.z = camera.position.z;
            // There is no component in CC which needs live updates when camera changes. Broadcasting an
            // onSettingsChanged Event would cause big performance issues
            // this.onSettingsChanged();
        }
    }

    /**
     * Broadcasts a settings-changed event with the new {Settings} object as a payload
     * @emits {settings-changed} on call
     */
    public onSettingsChanged() {

        this.settings.margin = this.computeMargin();

        if (this._lastDeltaState && this._settings.mode != KindOfMap.Delta) {
            this._lastDeltaState = false;
            this.onDeactivateDeltas();
        } else if (!this._lastDeltaState && this._settings.mode == KindOfMap.Delta) {
            this._lastDeltaState = true;
            this.onActivateDeltas();
        }

        this.$rootScope.$broadcast("settings-changed", this._settings);
    }

    /**
     * updates the settings object according to url parameters. url parameters are named like the accessors of the Settings object. E.g. scale.x or areaMetric
     * @emits {settings-changed} transitively on call
     */
    public updateSettingsFromUrl() {

        let ctx = this;

        let iterateProperties = function (obj, prefix) {
            for (let i in obj) {
                if (obj.hasOwnProperty(i) && i !== "map" && i) {

                    if (typeof obj[i] === "string" || obj[i] instanceof String) {
                        //do not iterate over strings
                    } else {
                        iterateProperties(obj[i], i + ".");
                    }

                    const res = ctx.urlService.getParam(prefix + i);

                    if (res) {

                        if (res == "true") {
                            obj[i] = true;
                        } else if (res == "false") {
                            obj[i] = false;
                        } else if (res === 0 || res) {

                            let val = parseFloat(res);

                            if (isNaN(val)) {
                                val = res;
                            }

                            obj[i] = val;
                        } else {
                            obj[i] = res;
                        }

                        // if we work with strings here it can cause errors in other parts of the app, check with console.log(typeof obj[i], obj[i]);
                    }

                }
            }
        };

        iterateProperties(this._settings, "");


        this.onSettingsChanged();

    }

    /**
     * @returns {number}
     *
     * Function that computes the margin applied to a scenario related the square root of (the area divided
     * by the number of buildings)
     */
    public computeMargin(
        map: CodeMap = this.dataService.data.renderMap,
        areaMetric: string = this.settings.areaMetric,
        settingsMargin: number = this.settings.margin,
        dynamicMargin: boolean = this.settings.dynamicMargin
    ): number {

        let margin: number;
        if (map !== null && dynamicMargin) {
            let root: CodeMapNode = map.root;

            let leaves = hierarchy<CodeMapNode>(root).leaves();
            let numberOfBuildings = 0;
            let totalArea = 0;
            leaves.forEach((c: HierarchyNode<CodeMapNode>) => {
                numberOfBuildings++;
                if(c.data.attributes && c.data.attributes[areaMetric]){
                    totalArea += c.data.attributes[areaMetric];
                }
            });

            margin = SettingsService.MARGIN_FACTOR * Math.round(Math.sqrt(
                (totalArea / numberOfBuildings)));

            margin = Math.min(100,Math.max(SettingsService.MIN_MARGIN, margin));
        }

        else {
            margin = settingsMargin || SettingsService.MIN_MARGIN;
        }

        return margin;
    }

    /**
     * Updates query params to current settings
     */
    public getQueryParamString() {

        let result = "";

        let iterateProperties = function (obj, prefix) {
            for (let i in obj) {
                if (obj.hasOwnProperty(i) && i !== "map" && i) {

                    if (typeof obj[i] === "string" || obj[i] instanceof String) {
                        //do not iterate over strings
                    } else {
                        iterateProperties(obj[i], i + ".");
                    }

                    if (typeof obj[i] === "object" || obj[i] instanceof Object) {
                        //do not print objects in string
                    } else {
                        result += "&" + prefix + i + "=" + obj[i];
                    }

                }

            }

        };

        iterateProperties(this._settings, "");

        return "?" + result.substring(1);

    }

    /**
     * Avoids the excesive calling of updateSettings with standard settings in order to increase the efficiency
     * When the function is called with an argument it calls updateSettings in order to avoid the lost of the information
     * contained in that argument.
     *
     * @param {Settings} settings
     */
    public applySettings(settings?: Settings) {

        if (settings) {
            this.updateSettings(settings);
        }

        else {
            this.numberOfCalls++;
            if (this.numberOfCalls > 4) {
                this.numberOfCalls = 0;
                this.onSettingsChanged();
            }
            else {
                let currentCalls = this.numberOfCalls;
                let _this = this;

                setTimeout(function () {
                    if (currentCalls == _this.numberOfCalls) {
                        this.numberOfCalls = 0;
                        _this.onSettingsChanged();
                    }

                }, 400);
            }
        }
    }

    /**
     * Applies given settings. ignores map. this ensures to copy settings object and prevent side effects
     * @param {Settings} settings
     */
    private updateSettings(settings: Settings) {
        this._settings.neutralColorRange.to = settings.neutralColorRange.to;
        this._settings.neutralColorRange.from = settings.neutralColorRange.from;
        this._settings.neutralColorRange.flipped = settings.neutralColorRange.flipped;

        this._settings.camera.x = settings.camera.x;
        this._settings.camera.y = settings.camera.y;
        this._settings.camera.z = settings.camera.z;

        this._settings.scaling.x = settings.scaling.x;
        this._settings.scaling.y = settings.scaling.y;
        this._settings.scaling.z = settings.scaling.z;

        this._settings.areaMetric = settings.areaMetric + "";
        this._settings.colorMetric = settings.colorMetric + "";
        this._settings.heightMetric = settings.heightMetric + "";

        this._settings.amountOfTopLabels = settings.amountOfTopLabels;
        this._settings.margin = settings.margin;
        this._settings.mode = settings.mode;
        this._settings.operation = settings.operation;
        this._settings.deltaColorFlipped = settings.deltaColorFlipped;
        this._settings.enableEdgeArrows = settings.enableEdgeArrows;
        this._settings.hideFlatBuildings = settings.hideFlatBuildings;
        this._settings.maximizeDetailPanel = settings.maximizeDetailPanel;
        this._settings.invertHeight = settings.invertHeight;
        this._settings.dynamicMargin = settings.dynamicMargin;
        this._settings.isWhiteBackground = settings.isWhiteBackground;
        this._settings.blacklist = settings.blacklist;
        this._settings.focusedNodePath = settings.focusedNodePath;
        this._settings.searchedNodePaths = settings.searchedNodePaths;
        this._settings.searchPattern = settings.searchPattern;

        //TODO what to do with map ? should it even be a part of settings ? deep copy of map ?
        this._settings.map = settings.map || this.settings.map;

        this.onSettingsChanged();

    }

    //TODO return new copy ? this would need a change listener for angular...
    get settings(): Settings {
        return this._settings;
    }

    /**
     * Returns a metric from the metrics object. If it is not found the last possible metric will be returned.
     * @param id id
     * @param {object} metrics metrics object
     * @returns {string} metric
     */
    private getMetricByIdOrLast(id: number, metrics: string[]): string {
        return metrics[Math.min(id, metrics.length - 1)];
    }

    /**
     * corrects settings, if the chosen metric is not available in the current map, the first three metrics are chosen as a default.
     * @param {Settings} settings
     */
    private correctSettings(settings) {
        const result = settings;
        result.map = this._settings.map; //do not change the map
        result.areaMetric = this.getMetricOrDefault(this.dataService.data.metrics, settings.areaMetric, this.getMetricByIdOrLast(0, this.dataService.data.metrics));
        result.heightMetric = this.getMetricOrDefault(this.dataService.data.metrics, settings.heightMetric, this.getMetricByIdOrLast(1, this.dataService.data.metrics));
        result.colorMetric = this.getMetricOrDefault(this.dataService.data.metrics, settings.colorMetric, this.getMetricByIdOrLast(2, this.dataService.data.metrics));
        return result;
    }

    /**
     * Checks if the given metricName is in the metricsArray. If it is in there, we return it, else we return the defaultValue.
     * @param {String[]} metricsArray an array of metric names
     * @param {String} metricName a metric name to look for
     * @param {String} defaultValue a default name in case metricName was not found
     */
    private getMetricOrDefault(metricsArray, metricName, defaultValue) {
        let result = defaultValue;
        metricsArray.forEach((metric) => {
            if (metric + "" === metricName + "") {
                result = metric;
            }
        });

        return result;
    }


}