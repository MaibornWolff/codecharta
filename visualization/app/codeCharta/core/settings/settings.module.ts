import angular from "angular";

import "../url/url.module";
import "../data/data.module";
import "../statistic/statistic.module";
import "../../ui/codeMap/threeViewer/threeViewer";

import {SettingsService} from "./settings.service";

angular.module(
    "app.codeCharta.core.settings",
    [
        "app.codeCharta.core.url",
        "app.codeCharta.core.data",
        "app.codeCharta.ui.codeMap.threeViewer",
        "app.codeCharta.core.statistic"
    ]
)
    .service(
        SettingsService.SELECTOR, SettingsService
    );

