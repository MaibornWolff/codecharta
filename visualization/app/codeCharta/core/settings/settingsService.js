"use strict";

import {Settings} from "./model/settings";
import {Scale} from "./model/scale";
import {Range} from "./model/range";

/**
 * Stores and manipulates the current settings
 */ 
class SettingsService {

    /* ngInject */

    /**
     * @constructor 
     * @param {UrlService} urlService 
     * @param {DataService} dataService
     * @param {Scope} $rootScope 
     */
    constructor(urlService, dataService, $rootScope) {

        /**
         * @type {UrlService}
         */
        this.urlService = urlService;

        /**
         * @type {DataService}
         */
        this.dataService = dataService;

        /**
         * @type {Scope}
         */
        this.rootScope = $rootScope;

        let ctx = this;

       /**
        * @type {Settings}
        */
        this.settings = new Settings(
            dataService.data.currentmap,
            new Range(10,20,false),
            ctx.getMetricByIdOrLast(0, dataService.data.metrics),
            ctx.getMetricByIdOrLast(1, dataService.data.metrics),
            ctx.getMetricByIdOrLast(2, dataService.data.metrics),
            true,
            false,
            1,
            new Scale(1,1,1),
            1
        );



        $rootScope.$on("data-changed", (event,data) => {
           ctx.onDataChanged(data);
        });

    }

    /**
     * change the map and metric settings according to the parameter.
     * @listens {data-changed} called on data-changed
     * @emits {settings-changed} transitively on call
     * @param {DataModel} data
     */
    onDataChanged(data) {

        this.settings.map = data.currentmap;

        if(data.metrics.indexOf(this.settings.areaMetric) === -1){
            //area metric is not set or not in the new metrics and needs to be chosen
            this.settings.areaMetric = this.getMetricByIdOrLast(0, data.metrics);

        }

        if(data.metrics.indexOf(this.settings.heightMetric) === -1){
            //height metric is not set or not in the new metrics and needs to be chosen
            this.settings.heightMetric = this.getMetricByIdOrLast(1, data.metrics);
        }

        if(data.metrics.indexOf(this.settings.colorMetric) === -1){
            //color metric is not set or not in the new metrics and needs to be chosen
            this.settings.colorMetric = this.getMetricByIdOrLast(2, data.metrics);
        }

        this.onSettingsChanged();

    }

    /**
     * Broadcasts a settings-changed event with the new {Settings} object as a payload
     * @emits {settings-changed} on call
     */
    onSettingsChanged() {
        this.rootScope.$broadcast("settings-changed", this.settings);

    }

    /**
     * updates the settings object according to url parameters.
     * @emits {settings-changed} transitively on call
     */
    updateSettingsFromUrl() {

        //for every property in settings...
        for (const property in this.settings) {

            //...that is not inherited from object...
            if (this.settings.hasOwnProperty(property) && property !== "map") {

                //...get the query param value for the property...
                const val = this.urlService.getParam(property);

                if (val && val.length > 0) {

                    //if it exists, set ist the settings value
                    this.settings[property] = val;
                }
            }
        }

        //TODO map some special keys like neutralColorRange

        this.onSettingsChanged();

    }

    /**
     * Applies given settings
     * @param {Settings} settings
     */
    applySettings(settings) {
        this.settings.importSettingValues(this.correctSettings(settings));
        this.onSettingsChanged();
    }

    /**
     * Returns a metric from the metrics object. If it is not found the last possible metric will be returned.
     * @param id id
     * @param {object} metrics metrics object
     * @returns {string} metric
     */
    getMetricByIdOrLast (id, metrics) {
        return metrics[Math.min(id, metrics.length-1)];
    }

    /**
     * corrects settings, if the chosen metric is not available in the current map, the first three metrics are chosen as a default.
     * @param {Settings} settings
     */
    correctSettings(settings){
        const result = settings;
        result.areaMetric = this.getMetricOrDefault(this.dataService.data.metrics, settings.areaMetric,this.getMetricByIdOrLast(0, this.dataService.data.metrics));
        result.heightMetric = this.getMetricOrDefault(this.dataService.data.metrics, settings.heightMetric, this.getMetricByIdOrLast(1, this.dataService.data.metrics));
        result.colorMetric = this.getMetricOrDefault(this.dataService.data.metrics, settings.colorMetric, this.getMetricByIdOrLast(2, this.dataService.data.metrics));
        return result;
    }

    /**
      * Checks if the given metricName is in the metricsArray. If it is in there, we return it, else we return the defaultValue.
      * @param {String[]} metricsArray an array of metric names
      * @param {String} metricName a metric name to look for
      * @param {String} defaultValue a default name in case metricName was not found
      */
    getMetricOrDefault(metricsArray, metricName, defaultValue) {
        let result = defaultValue;
        metricsArray.forEach((metric) => {
            if(metric + "" === metricName + ""){
                    result = metric;
                }
        });

        return result;
    }
}

export {SettingsService};