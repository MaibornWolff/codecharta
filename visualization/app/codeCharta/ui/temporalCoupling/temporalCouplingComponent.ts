import {SettingsServiceSubscriber, SettingsService, Settings} from "../../core/settings/settings.service";
import {ITimeoutService} from "angular";
import {CodeMap, CodeMapDependency, CodeMapNode} from "../../core/data/model/CodeMap";
import {hierarchy} from "d3-hierarchy";

export class TemporalCouplingController implements SettingsServiceSubscriber {

    public  temporalCoupling: CodeMapDependency[] = null;

    /* @ngInject */
    constructor(private $timeout: ITimeoutService,
                private $scope,
                private settingsService: SettingsService) {

        this.onSettingsChanged(settingsService.settings);
        this.settingsService.subscribe(this);
    }

    onSettingsChanged(s: Settings) {
        this.updateTemporalCouplingDependencies(this.settingsService.settings.map);
    }

    public onResetDependencies() {
        var map = this.settingsService.settings.map;
        if (map && map.dependencies && map.dependencies.temporal_coupling) {

            for (var i = 0; i < this.settingsService.settings.map.dependencies.temporal_coupling.length; i++) {
                this.settingsService.settings.map.dependencies.temporal_coupling[i].visible = false;
            }
        }
        this.settingsService.applySettings();
    }

    onClickCouple(couple: CodeMapDependency) {
        this.setNewVisibilityInSettingsService(this.settingsService.settings.map.dependencies.temporal_coupling, couple);
        this.settingsService.applySettings();
    }

    setNewVisibilityInSettingsService(deps: CodeMapDependency[], couple: CodeMapDependency) {
        let coupleIndex = deps.indexOf(couple);
        let oldVisibility = deps[coupleIndex].visible;
        let newVisibility = !(oldVisibility === true);
        this.settingsService.settings.map.dependencies.temporal_coupling[coupleIndex].visible = newVisibility;
    }

    private updateTemporalCouplingDependencies(map: CodeMap) {
        if (map && map.dependencies && map.dependencies.temporal_coupling) {
            this.temporalCoupling = map.dependencies.temporal_coupling;

            for (var couple of this.temporalCoupling) {
                if (couple.visible !== true) {
                    couple.visible = false;
                }
            }

            this.temporalCoupling = this.temporalCoupling.filter(
                couple =>
                    couple.averageRevs >= this.settingsService.settings.minimumAverageRevs &&
                    this.isCoupleVisibleInCodeMap(couple, map)
            );

            if (this.settingsService.settings.intelligentTemporalCouplingFilter) {
                this.temporalCoupling = this.temporalCoupling.filter(
                    couple => this.isEligibleCouple(couple)
                );
            }
        }
    }

    private isCoupleVisibleInCodeMap(couple, map) {
        var isNodeInCodeMap = false;
        var isDependantNodeInCodeMap = false;

        hierarchy<CodeMapNode>(map.root).leaves().forEach(function (node) {

            if (node.data.path == couple.node && node.data.visible) {
                isNodeInCodeMap = true;
            }
            if (node.data.path == couple.dependantNode && node.data.visible) {
                isDependantNodeInCodeMap = true;
            }
        });
        return isNodeInCodeMap && isDependantNodeInCodeMap;
    }

    private isEligibleCouple(couple) {

        let nodenameBlacklist = [
            /package-lock\.json/i,
            /package\.json/i,
            /CHANGELOG.*/i,
            /README.*/i,
            /.*\.log/i,
            /gradle\.properties/i,
            /build\.gradle/i,
            /sample1 only leaf/i,
        ];
        let isEligibleCouple = true;

        nodenameBlacklist.forEach(function (re) {

            let isNodeNameMatch = couple.nodeFilename.search(re) != -1;
            let isDependantNodeNameMatch = couple.dependantNodeFilename.search(re) != -1;

            if (isNodeNameMatch || isDependantNodeNameMatch) {
                isEligibleCouple = false
            }
        });
        return isEligibleCouple;
    }
}

export const temporalCouplingComponent = {
    selector: "temporalCouplingComponent",
    template: require("./temporalCoupling.html"),
    controller: TemporalCouplingController
};




