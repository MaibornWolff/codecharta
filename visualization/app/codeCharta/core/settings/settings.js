"use strict";

import "../url/url.js";
import "../data/data.js";

import {SettingsService} from "./settingsService.js";

angular.module(
    "app.codeCharta.core.settings",
    ["app.codeCharta.core.url", "app.codeCharta.core.data"]
);

angular.module("app.codeCharta.core.settings").service(
    "settingsService", SettingsService
);

