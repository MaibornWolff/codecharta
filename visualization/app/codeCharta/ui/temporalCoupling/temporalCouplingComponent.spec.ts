import {ScenarioButtonsController} from "./scenarioButtonsComponent";
import {ITimeoutService} from "angular";
import {SettingsService} from "../../core/settings/settings.service";
import {TemporalCouplingController} from "./temporalCouplingComponent";
import {CodeMapDependency} from "../../core/data/model/CodeMap";

describe("TemporalCouplingComponent", () => {

    let temporalCouplingController: TemporalCouplingController;
    let $timeout: ITimeoutService;
    let $scope;
    let settingsService: SettingsService;

    const couple: CodeMapDependency = {
        node: "/root/Anode",
        nodeFilename: "Anode",
        dependantNode: "/root/AnotherNode",
        dependantNodeFilename: "AnotherNode",
        pairingRate: 42,
        averageRevs: 21,
        visible: true,
    };

    function rebuildSUT() {
        temporalCouplingController = new TemporalCouplingController($timeout, $scope, settingsService);
    }

    function mockEverything() {

        const ITimeoutServiceMock = jest.fn<ITimeoutService>(() => ({}));

        $timeout = new ITimeoutServiceMock();

        const $scopeMpck = jest.fn(() => ({}));

        $scope = new $scopeMpck();

        const SettingsService = jest.fn<SettingsService>(() => ({
            settings: {
                map: {
                    dependencies: {
                        temporal_coupling: [
                            {
                                node: "/root/Anode",
                                nodeFilename: "Anode",
                                dependantNode: "/root/AnotherNode",
                                dependantNodeFilename: "AnotherNode",
                                pairingRate: 42,
                                averageRevs: 21,
                                visible: true,
                            },
                            {
                                node: "/root/parent/Anode",
                                nodeFilename: "Anode",
                                dependantNode: "/root/parent/AnotherNode",
                                dependantNodeFilename: "AnotherNode",
                                pairingRate: 42,
                                averageRevs: 21,
                                visible: true,
                            },
                        ]
                    },
                },
                minimumAverageRevs: 15,
            },
            subscribe: jest.fn(),
            applySettings: jest.fn(),
        }));

        settingsService = new SettingsService();

        rebuildSUT();
    }

    beforeEach(()=>{
        mockEverything();
    });


    it("should toggle visibility for clicked couple", () => {
        const couple = {
            node: "/root/Anode",
            nodeFilename: "Anode",
            dependantNode: "/root/AnotherNode",
            dependantNodeFilename: "AnotherNode",
            pairingRate: 42,
            averageRevs: 21,
            visible: true
        };
        settingsService.settings.map.dependencies.temporal_coupling[0] = couple;
        temporalCouplingController.onClickCouple(couple);
        expect(settingsService.settings.map.dependencies.temporal_coupling).toMatchSnapshot();
    });

    it("should reset visibility for all dependencies to false", () => {
        temporalCouplingController.onResetDependencies();
        expect(settingsService.settings.map.dependencies.temporal_coupling).toMatchSnapshot();
    });

    it("should be eligible Couple", () => {
        expect(temporalCouplingController.isEligibleCouple(couple)).toBe(true);
    });

    it("should be eligible Couple", () => {
        couple.node = "/root/package-lock.json";
        couple.nodeFilename = "package-lock.json";
        expect(temporalCouplingController.isEligibleCouple(couple)).toBe(false);
    });

    it("should use filter when updating temporalCOuplingController.temporalCoupling", () => {
        settingsService.settings.map.dependencies.temporal_coupling[0] = {
            node: "/root/parent/children/package.json",
            nodeFilename: "package.json",
            dependantNode: "/root/AnotherNode",
            dependantNodeFilename: "AnotherNode",
            pairingRate: 42,
            averageRevs: 10,
            visible: false
        };
        settingsService.settings.intelligentTemporalCouplingFilter = true;
        temporalCouplingController.updateTemporalCouplingDependencies(settingsService.settings.map);
        expect(temporalCouplingController.temporalCoupling.length).toEqual(1);
    });



});