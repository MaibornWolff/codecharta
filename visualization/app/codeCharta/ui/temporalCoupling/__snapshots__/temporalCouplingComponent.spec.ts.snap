// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemporalCouplingComponent should reset visibility for all dependencies to false 1`] = `
Array [
  Object {
    "averageRevs": 21,
    "dependantNode": "/root/AnotherNode",
    "dependantNodeFilename": "AnotherNode",
    "node": "/root/Anode",
    "nodeFilename": "Anode",
    "pairingRate": 42,
    "visible": false,
  },
  Object {
    "averageRevs": 21,
    "dependantNode": "/root/parent/AnotherNode",
    "dependantNodeFilename": "AnotherNode",
    "node": "/root/parent/Anode",
    "nodeFilename": "Anode",
    "pairingRate": 42,
    "visible": false,
  },
]
`;

exports[`TemporalCouplingComponent should toggle visibility for clicked couple 1`] = `
Array [
  Object {
    "averageRevs": 21,
    "dependantNode": "/root/AnotherNode",
    "dependantNodeFilename": "AnotherNode",
    "node": "/root/Anode",
    "nodeFilename": "Anode",
    "pairingRate": 42,
    "visible": false,
  },
  Object {
    "averageRevs": 21,
    "dependantNode": "/root/parent/AnotherNode",
    "dependantNodeFilename": "AnotherNode",
    "node": "/root/parent/Anode",
    "nodeFilename": "Anode",
    "pairingRate": 42,
    "visible": true,
  },
]
`;
